File: challenge1/Java/FindBob.java
import java.io.*;
import java.util.*;

public class Solution {
    
    private static String canFindBob(String[][] grid){
        int[] alicePos = findAlicePos(grid);
        return canFindBob(alicePos[0], alicePos[1], grid) ? "YES" : "NO";
    }
    
    private static int[] findAlicePos(String[][] grid){
        for(int i = 0; i < grid.length; i++){
            for (int j = 0; j < grid[i].length; j++){
                if ("A".equals(grid[i][j])){
                    return new int[]{i, j};
                }
            }
        }
        
        return new int[]{-1, -1};
    }
    
    private static boolean canFindBob(int r, int c, String[][] grid){
        if (r < 0 || r >= grid.length){
            return false;
        }
        
        if (c < 0 || c >= grid[r].length){
            return false;
        }
        
        if ("#".equals(grid[r][c])){
            return false;
        }
        
        if ("B".equals(grid[r][c])){
            return true;
        }
        
        return canFindBob(r + 1, c, grid) || canFindBob(r, c + 1, grid);
    }
    
    public static void main(String[] args) throws IOException {
        /* Enter your code here. Read input from STDIN. Print output to STDOUT. Your class should be named Solution. */
        
        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        //n = rows, m = cols
        String[] firstLineInput = bufferedReader.readLine().replaceAll("\\s+$", "").split(" ");
        int n = Integer.parseInt(firstLineInput[0]);
        int m = Integer.parseInt(firstLineInput[1]);
        
        String[][] grid = new String[n][m];
        for(int i = 0; i < n; i++){
            grid[i] = bufferedReader.readLine().replaceAll("\\s+$", "").split("");
        }
        
        String result = canFindBob(grid);
        
        bufferedWriter.write(String.valueOf(result));
        bufferedWriter.newLine();

        bufferedReader.close();
        bufferedWriter.close();

    }
}

File: challenge1/Python/findBob.py
def dfs(x, y):
    global n, m, grid, found, vis

    if x >= n or y >= m:
        return
    if grid[x][y] == '#':
        return
    if vis[x][y] == True:
        return
    vis[x][y] = True
    if grid[x][y] == 'B':
        found = True
        return
    dfs(x + 1, y)
    dfs(x, y + 1)


def main():
    global n, m, grid, found, vis
    found = False
    vis = []
    (n, m) = map(int, input().split())
    grid = []
    for i in range(n):
        s = input()
        grid.append(s)
        vis.append([False] * m)
    (x, y) = (-1, -1)
    for i in range(0, n):
        for j in range(0, m):
            if grid[i][j] == 'A':
                (x, y) = (i, j)
    dfs(x, y)
    if found == True:
        print ('YES')
    else:
        print ('NO')

if __name__ == '__main__':
    main()